controllers/openshift_controller.go

```diff
@@ -1679,6 +1679,69 @@ func (r *KataConfigOpenShiftReconciler) getMcpByName(mcpName string) (*mcfgv1.Ma
 	return mcp, nil
 }
 
+func (r *KataConfigOpenShiftReconciler) processDoneNode(node *corev1.Node) error {
+
+	isConvergedCluster, err := r.checkConvergedCluster()
+	if err != nil {
+		return err
+	}
+
+	targetMcpName := func() string {
+		if isConvergedCluster {
+			return "master"
+		}
+		_, nodeLabeledForKata := node.Labels["node-role.kubernetes.io/kata-oc"]
+		if nodeLabeledForKata {
+			return "kata-oc"
+		} else {
+			return "worker"
+		}
+	}()
+
+	targetMcp, err := r.getMcpByName(targetMcpName)
+	if err != nil {
+		return err
+	}
+
+	installCompletedList := &r.kataConfig.Status.InstallationStatus.Completed.CompletedNodesList
+	uninstallCompletedList := &r.kataConfig.Status.UnInstallationStatus.Completed.CompletedNodesList
+
+	currentNodeConfig, ok := node.Annotations["machineconfiguration.openshift.io/currentConfig"]
+
+	log := r.Log.WithName("updateStatus").WithValues("node name", node.GetName(), "target pool", targetMcpName, "current config", currentNodeConfig, "target pool config", targetMcp.Spec.Configuration.Name)
+	if ok && currentNodeConfig == targetMcp.Spec.Configuration.Name {
+		// This Node is at its target configuration so it's going to
+		// go to a Completed list, either installation's or uninstallation's.
+		// To decide which we consider:
+		// - if uninstallation is in progress, any settled Node must
+		//   belong to uninstallation's Complete list as no
+		//   installations can take place while uninstalling kata from
+		//   cluster
+		// - OTOH if installation is in progress, settled Nodes will go
+		//   to installation's Completed list by default as expected
+		//   *unless* the Node's target pool is "worker" in which case
+		//   the Node shouldn't and doesn't have kata on it and thus
+		//   belongs to uninstallation's Completed list.
+		if r.kataConfig.Status.InstallationStatus.IsInProgress == corev1.ConditionTrue {
+			if targetMcpName == "worker" {
+				log.Info("putting done node on uninstallation completed list")
+				*uninstallCompletedList = append(*uninstallCompletedList, node.GetName())
+				r.kataConfig.Status.UnInstallationStatus.Completed.CompletedNodesCount = int(targetMcp.Status.UpdatedMachineCount)
+			} else {
+				log.Info("putting done node on installation completed list")
+				*installCompletedList = append(*installCompletedList, node.GetName())
+				r.kataConfig.Status.InstallationStatus.Completed.CompletedNodesCount = int(targetMcp.Status.UpdatedMachineCount)
+			}
+		} else if r.kataConfig.Status.UnInstallationStatus.InProgress.IsInProgress == corev1.ConditionTrue {
+			log.Info("putting done node on uninstallation completed list")
```

Comment:
It could be valuable to distinguish this log from the other one above right away. Maybe make the first one "putting done worker node" ?

Reply:
Yes, I understand where you're coming from, I was trying to find a good way to do that. I gave up since the philosophy of the logger used here is to use the key/value pairs to convey additional information rather than embed it in the message itself printf()-style, and I didn't feel comfortable going against it.

Note that although the two messages look the same in the source code they aren't the same actually since we're logging here via a custom logger with added keys/values - so the information that this is a member of "worker" is there already .

But I'm not 100% happy with the result either so if the difference doesn't seem enough I'm open to change this.

Reply:
The logger is :

```
r.Log.WithName("updateStatus").WithValues("node name", node.GetName(), "target pool", targetMcpName, "current config", currentNodeConfig, "target pool config", targetMcp.Spec.Configuration.Name)
```

In a non-converged scenario, `targetMcpName` will be `worker` in both cases. Only differentiator that comes to mind is the source MCP, either `kata-oc` (case 2) or `worker` (case 1).

In a converged scenario, the case 1 cannot be reached since "worker" will never be the target MCP and case 2 is `master`.

It isn't needed to determine the source MCP. Mentioning "worker" (as originating MCP) in case 1 seems to be simple enough.

Reply:
I'm not sure if I'm following the argument correctly.  Let me recap the main aspects of the situation as I understand them to see if there are any problems with it:
- this code deals with Nodes that are Done and whose MachineConfig matches the MachineConfig of the MCP that they belong to. In other words, these Nodes are currently settled and not moving anywhere. So I'm not sure what you mean by "source MCP".
- because of the previous point, these Nodes go to a Completed list.  Both of the places you point out deal with Nodes that go to the uninstallation's Completed list.  A settled Node can end up on that list for one of two different reasons.
- one is because we're uninstalling kata from a cluster.  A settled Node in that situation can't have kata installed on it, kata has either been uninstalled already or it was never installed on the Node in the first place.  This the case 2.
- the other is because we're updating kata installation on a cluster (which is handled as installation internally) and part of that update was removing kata from a Node.  This is the case 1.  Such Node is identified by belonging to the "worker" MCP.
Note that this code also handles Nodes that belong to "worker" either because they never had kata installed on them or because they did but it was uninstalled previously already, since all these scenarios are indistinguishable I believe.

We simply see a Node that should and does belong to the "worker" MCP and we have no idea how it got there.

(It works the same on converged clusters but it's a bit simpler - since no Node can belong to "worker" the logic simplifies to "if we're putting kata on cluster any settled Node must have completed its kata installation, if we're removing kata from cluster any settled Node must have completed its kata uninstallation".)

Reply:
My point is that the two instances of `Log.Info()` produce exactly the same output (I've verified it) => it is impossible to guess which branch the logs comes from with a simple grep.

Reply:
Okay, that's true, but if there's a difference between the two cases, it's not that the Node is in "worker" (which is the case in both log messages) but that we're "installing" in case 1 (in the confusingly wide meaning of the word) but uninstalling in case 2.  So do you want me to include _that_ information in the log messages?

Reply:
Hm, but what does the proposed code do when `InstallationStatus.IsInProgress == corev1.ConditionTrue` and `targetMcpName == "worker"`?  Correct me if I'm wrong but it seems to me that it will (incorrectly) do nothing, considering that `InstallationStatus.IsInProgress == corev1.ConditionTrue` implies `UnInstallationStatus.InProgress.IsInProgress == corev1.ConditionFalse`.

I guess this could be fixed, perhaps by changing the `else if` condition to `UnInstallationStatus.InProgress.IsInProgress == corev1.ConditionTrue || targetMcpName == "worker"`, at which point however I'd strongly doubt that the new formulation is any better, in fact it seems to me quite a lot more tricky to understand and verify.

Reply:
Alright, maybe disentangle the checking and the processing then ?

```
		installingNode  := false
		uninstallingNode := false
		if r.kataConfig.Status.InstallationStatus.IsInProgress == corev1.ConditionTrue {
			if targetMcpName == "worker" {
				uninstallingNode = true
			} else {
				installingNode = true
			}
		} else if r.kataConfig.Status.UnInstallationStatus.InProgress.IsInProgress == corev1.ConditionTrue {
			uninstallingNode = true
		}

		if installingNode {
			log.Info("putting done node on installation completed list")
			*installCompletedList = append(*installCompletedList, node.GetName())
			r.kataConfig.Status.InstallationStatus.Completed.CompletedNodesCount = int(targetMcp.Status.UpdatedMachineCount)
		} else if uninstallingNode {
			log.Info("putting done node on uninstallation completed list")
			*uninstallCompletedList = append(*uninstallCompletedList, node.GetName())
			r.kataConfig.Status.UnInstallationStatus.Completed.CompletedNodesCount = int(targetMcp.Status.UpdatedMachineCount)
		}
```

Reply:
Yes you're right. Getting rid of the duplication isn't interesting here.